legate/numpy/runtime.py

```diff
@@ -642,72 +628,72 @@ def find_or_create_key_partition(self):
+                # We're invertible, so we can reuse the key partition of the
+                # root, and specifically the subset of root tiles that cover
+                # the view.
+                # Construct the transformations from view to root space
+                v2r_permute = self.transform.trans
+                # v2r_elem = self.transform
+                v2r_tile = AffineTransform(
+                    len(root.shape), len(self.shape), False
+                )
+                v2r_tile.trans = v2r_permute
+                v2r_tile.offset = color_lo
+                # Construct the transformations from root to view space
+                r2v_permute = self.transform.trans.T
+                r2v_elem = AffineTransform(
+                    len(self.shape), len(root.shape), False
+                )
+                r2v_elem.trans = r2v_permute
+                r2v_elem.offset = -(r2v_permute @ self.transform.offset)
+                r2v_tile = AffineTransform(
+                    len(self.shape), len(root.shape), False
+                )
+                r2v_tile.trans = r2v_permute
+                r2v_tile.offset = -(r2v_permute @ color_lo)
+                # Sanity check
+                assert r2v_elem.apply(lo) == (0,) * len(self.shape)
+                assert r2v_elem.apply(hi) == tuple(x - 1 for x in self.shape)
                 # We need to make a make a special sharding functor here that
                 # projects the points in our launch space back into the space
-                # of the root partitions sharding space
-                # First construct the affine mapping for points in our launch
-                # space back into the launch space of the root
+                # of the root partition's sharding space.
                 # This is the special case where we have a special shard
                 # function and sharding space that is different than our normal
                 # launch space because it's a subset of the root's launch space
-                launch_transform = AffineTransform(
-                    len(root.shape), len(self.shape), False
-                )
-                launch_transform.trans = self.transform.trans
-                launch_transform.offset = color_lo
                 self.shard_function = (
                     self.runtime.find_or_create_transform_sharding_functor(
-                        launch_transform
+                        v2r_tile
                     )
                 )
-                tile_offset = np.zeros((len(self.shape),), dtype=np.int64)
-                for n in range(len(self.shape)):
-                    nonzero = False
-                    for m in range(len(root.shape)):
-                        if inverse[n, m] == 0:
-                            continue
-                        nonzero = True
-                        break
-                    if not nonzero:
-                        tile_offset[n] = 1
```

Comment:
I couldn't understand what tile_offset was doing.

Reply:
See the case above. I think I was trying to handle the case where dimensions of size 1 were added or removed as part of the creation of a view.

Reply:
You are right, the variable name confused me. It looks like it's making sure the view tile shape is at least 1-wide on every dimension, for cases where the view adds a dimension with `newaxis`.

```
tile_offset[n] = 1 if (forall m: inverse[n,m] == 0) else 0
color_tile = tuple( (inverse @ color_tile).flatten() + tile_offset )
self.key_partition = _find_or_create_partition(tile_shape=color_tile)
```